src/main/java/heronarts/lx/buffer/FloatTensorAddBlend.java

```diff
@@ -0,0 +1,161 @@
+    outG.muli(255.0).addi(srcG255.mul(effectiveAlpha).div(256.0));
+    outB.muli(255.0).addi(srcB255.mul(effectiveAlpha).div(256.0));
+    outAlpha.muli(255.0).addi(effectiveAlpha);
+
+    // Clip to [0, 255]
+    outR.assign(Transforms.min(outR, 255.0));
+    outG.assign(Transforms.min(outG, 255.0));
+    outB.assign(Transforms.min(outB, 255.0));
+    outAlpha.assign(Transforms.min(outAlpha, 255.0));
+
+    // Convert back to [0,1] range
+    outR.divi(255.0);
+    outG.divi(255.0);
+    outB.divi(255.0);
+    outAlpha.divi(255.0);
+
+//    // Clip to valid range using putWhere (most efficient)
+//    INDArray maxVal = Nd4j.scalar(255.0);
+//    outR.putWhere(outR.gt(255.0), maxVal);
+//    outG.putWhere(outG.gt(255.0), maxVal);
+//    outB.putWhere(outB.gt(255.0), maxVal);
+//    outAlpha.putWhere(outAlpha.gt(255.0), maxVal);
+  }
+
+  /**
+   * @param mask Boolean tensor [N] or [N, 1] - true where blending should occur
+   */
+  @Override
+  public void blend(INDArray dst, INDArray src, double alpha, INDArray output, INDArray mask) {
+    validateInputs(DataType.FLOAT, dst, src, output);
+    validateMask(mask, dst.size(0));
+
+    if (output != dst) {
+      output.assign(dst);
+    }
+
+    // Ensure mask is the right shape for broadcasting - should be [N] not [N, 1]
+    INDArray broadcastMask = mask;
+    if (mask.rank() == 2) {
+      broadcastMask = mask.getColumn(0); // Convert [N, 1] to [N]
+    }
+
+    // Calculate the blend for all pixels
+    INDArray srcAlpha = src.getColumn(0);
+    INDArray srcR = src.getColumn(1);
+    INDArray srcG = src.getColumn(2);
+    INDArray srcB = src.getColumn(3);
+
+    // Apply exact integer formula (same as before)
+    INDArray srcAlpha255 = srcAlpha.mul(255.0);
+    INDArray srcR255 = srcR.mul(255.0);
+    INDArray srcG255 = srcG.mul(255.0);
+    INDArray srcB255 = srcB.mul(255.0);
+
+    double alphaScale = alpha * 256.0;
+    INDArray effectiveAlpha = srcAlpha255.mul(alphaScale).div(256.0);
+    INDArray roundingMask = effectiveAlpha.gte(127.5);
+    effectiveAlpha.addi(roundingMask.castTo(DataType.FLOAT));
+
+    // Calculate deltas (what to add to dst)
+    INDArray deltaR = srcR255.mul(effectiveAlpha).div(256.0);
+    INDArray deltaG = srcG255.mul(effectiveAlpha).div(256.0);
+    INDArray deltaB = srcB255.mul(effectiveAlpha).div(256.0);
+    INDArray deltaAlpha = effectiveAlpha;
+
+    // Apply mask: only add deltas where mask is true
+    // Use .mul() instead of .muli() to avoid in-place shape issues
+    deltaR = deltaR.mul(broadcastMask);
+    deltaG = deltaG.mul(broadcastMask);
+    deltaB = deltaB.mul(broadcastMask);
+    deltaAlpha = deltaAlpha.mul(broadcastMask);
+
+    // Apply to output (convert to 255 space, add deltas, clip, convert back)
+    INDArray outAlpha = output.getColumn(0);
+    INDArray outR = output.getColumn(1);
+    INDArray outG = output.getColumn(2);
+    INDArray outB = output.getColumn(3);
+
+    // Scale to 255, add deltas, clip, scale back
```

Comment:
Same comment here, don't think an ideal float implementation would do all this 255-scaling.